lib/Test/MockFile.pm

```diff
@@ -307,17 +466,74 @@ sub _strict_mode_violation {
     my @stack = _get_stack();
     return unless scalar @stack;    # skip the package
 
+    my $filename;
+
     # check it later so we give priority to authorized_strict_mode_packages
     my $file_arg = file_arg_position_for_command( $command, $at_under_ref );
 
-    my $filename = scalar @$at_under_ref <= $file_arg ? '<not specified>' : $at_under_ref->[$file_arg];
+    if ( $file_arg >= 0 ) {
+        $filename = scalar @$at_under_ref <= $file_arg ? '<not specified>' : $at_under_ref->[$file_arg];
+    }
 
     # Ignore stats on STDIN, STDOUT, STDERR
-    return if $filename =~ m/^\*?(?:main::)?[<*&+>]*STD(?:OUT|IN|ERR)$/;
+    return if defined $filename && $filename =~ m/^\*?(?:main::)?[<*&+>]*STD(?:OUT|IN|ERR)$/;
+
+    my $path = _abs_path_to_file($filename);
+
+    my $context = {
+        command      => $command,
+        filename     => $path,
+        at_under_ref => $at_under_ref
+    };    # object
+
+    return if _dispatch_strict_rules($context);
+
+    croak("Unknown strict mode violation for $command") if $file_arg == -1;
 
     confess("Use of $command to access unmocked file or directory '$filename' in strict mode at $stack[1] line $stack[2]");
 }
 
+sub _dispatch_strict_rules {
+    my ($context) = @_;
+
+    # rules dispatch
+    foreach my $rule (@STRICT_RULES) {
+
+        # validate file_rule
+        next unless _check_rule_condition->( $context->{filename}, $rule->{file_rule} );
+
+        # validate command_rule
+        next unless _check_rule_condition->( $context->{command}, $rule->{command_rule} );
+
+        my $answer = ref $rule->{action} ? $rule->{action}->($context) : $rule->{action};
+
+        # for now the only known answer to the rules, otherwise keep going
+        return 1 if $answer && $answer == 1;
+    }
+
+    return;
+}
+
+sub _check_rule_condition {
+    my ( $str, $match ) = @_;
+
+    return 1 unless defined $match;    # no restrictions
+    return   unless defined $str;      #
+
+    if ( my $isa = ref $match ) {
+        return 1 if $isa eq 'Regexp' && $str =~ $match;
+        if ( $isa eq 'ARRAY' ) {
+            foreach my $m (@$match) {
+                return 1 if _check_rule_condition( $str, $m );
+            }
+        }
+    }
+
+    return 1 if $str eq $match;
+
+    return;
```

Comment:
I think it's better if the rule itself describes what happens, but this would break simple rules. (We could try to still make them work.)

```perl
add_strict_rule(
    '*',
    '*',
    sub ($keyword, $file, $ctx ) {
        # We are only looking for keywords starting with "open"
        # Stop this rule, but keep looking for other rules
        $keyword =~ /^open/xms or return; # undef

        # From here on, allow all Temp::File checks
        # This passes stricture, don't check more rules
        $ctx->{'package'} eq 'Temp::File' and return 1;

        # I'm sorry, we reached an open keyword not used by Temp::File
        # so report this as a stricture failure
        return 0;
    },
);
```

Another example for the first scenario (which is missing in the PR):

```perl
my @interrupted_keyword_calls;

add_strict_rule(
    '*',
    '*',
    sub ($keyword, $file, $ctx ) {
        # Track all usages of calls 
        push @interrupted_keyword_calls, {
            'keyword'  => $keyword,
            'filename' => $keyword,
            'context'  => $ctx,
        };
    },
);
```